skbio/alignment/_path.py

```diff
@@ -0,0 +1,305 @@
+        # structure.
+        byte_arr = np.stack([x._bytes for x in msa._seqs])
+
+        # Get gap character code.
+        gap_chars = [ord(x) for x in msa.dtype.gap_chars]
+
+        # Identify gap positions, and convert them into a bit array, then create an
+        # alignment path based on it.
+        return cls.from_bits(np.isin(byte_arr, gap_chars))
+
+    def to_indices(self, gap=-1):
+        r"""Generate an array of indices of characters in the original sequences.
+
+        gap = "del": delete columns that have any gap, "mask": mask gaps, others
+        (default: -1):
+        fill gaps with this value.
+        """
+        valid_gaps = {-1, "del", "mask"}
+        if gap not in valid_gaps:
+            raise ValueError("Gap must be -1, 'del', or 'mask'.")
+
+        bits = np.squeeze(self.to_bits())
+        # TODO: Consider optimization using np.arange.
+        # thought: initiate [-1, -1, -1 ... -1], then add slices of arange into it
+        pos = np.repeat(1 - bits, self.lengths, axis=1)
+        idx = np.cumsum(pos, axis=1, dtype=int) - 1
+        if gap == "del":
+            keep = np.repeat(self.states == 0, self.lengths)
+            return idx[:, keep]
+        elif gap == "mask":
+            return np.ma.array(idx, mask=(1 - pos))
+        else:
+            return np.where(pos, idx, gap)
+
+    @classonlymethod
+    def from_indices(cls, indices, gap=-1):
+        r"""Create an alignment path from character indices in the original sequences.
+
+        gap can be a value or "mask", but cannot be "del".
+        """
+        if gap == "mask":
+            return cls.from_bits(np.ma.getmask(indices))
+        else:
+            return cls.from_bits(indices == gap)
+
+    def to_coordinates(self):
+        r"""Generate an array of segment coordinates in the original sequences."""
+        lens = self.lengths * (1 - self.to_bits())
+        col0 = np.zeros((self.shape[0], 1), dtype=int)
+        lens = np.append(col0, lens, axis=1)
+        return lens.cumsum(axis=1)
+
+    @classonlymethod
+    def from_coordinates(cls, coords):
+        r"""Generate the an alignment path from an array of segment coordinates."""
+        diff = np.diff(coords)
+        bits = diff == 0
+        lens = diff[bits.argmin(axis=0), np.arange(diff.shape[1])]
+        ints = np.packbits(bits, axis=0, bitorder="little")
+        if ints.shape[0] == 1:
+            ints = np.squeeze(ints, axis=0)
+        return cls(lens, ints, np.zeros(diff.shape[0], dtype=int))
+
+
+class PairAlignPath(AlignPath):
+    @overrides(AlignPath)
+    @classonlymethod
+    def from_bits(cls, bits):
+        r"""Create an alignment path from a bit array."""
+        # This should be faster than the generic solution I guess.
+        # TODO: Pending benchmark and optimization.
+
+        # Ensure bits is a 2D numpy array of ones and zeros.
+        if not isinstance(bits, np.ndarray) or bits.ndim != 2 or bits.shape[1] == 0:
```

Comment:
`bits` may be array-like (e.g., a NumPy array or a Python list), and one can use `np.atleast_2d` to convert it into a 2-D array.